### PR TITLE
feat(snap): add Snap compression option

### PR DESCRIPTION
**Pull request by om26er**
Fixes https://github.com/electron-userland/electron-builder/issues/5881

Signed-off-by: Omer Akram <om26er@gmail.com>

### CODE REVIEW

**Comment by om26er** [comment]
Did more investigation and found the above change won't work. electron builder uses `app-builder` to create the snap archive, which currently has `xz` compression hard-coded. https://github.com/develar/app-builder/blob/4dbd42fea73c5d6526009de9d077375c46c0d41a/pkg/package-format/snap/snap.go#L288

Will first with with `app-builder` upstream to optionally support LZO compression for snap

**Comment by om26er** [comment]
Just a heads up, the fix for app-builder is now up https://github.com/develar/app-builder/pull/65 -- Will likely need to adapt this PR once app-builder change lands (or may even have to wait for a new release of it ?)

**Comment by develar** [comment]
Thanks! I published ` app-builder-bin@4.0.0`. The reason for breaking change — Go 1.17 requires macOS 10.13 High Sierra or later. I decided that it is ok.

**Comment by om26er** [comment]
Sorry for the delay, I will get back to this PR during the weekend.

**Comment by om26er** [comment]
Now hit by https://github.com/electron-userland/electron-builder-binaries/issues/45

**Comment by ppd** [comment]
This PR works for me with app-builder-bin 4.1.0 (latest) and the missing hookup of the compression option:

```patch
diff --git a/packages/app-builder-lib/src/targets/snap.ts b/packages/app-builder-lib/src/targets/snap.ts
index 7f1dbd72..e6e0dbc9 100644
--- a/packages/app-builder-lib/src/targets/snap.ts
+++ b/packages/app-builder-lib/src/targets/snap.ts
@@ -222,6 +222,10 @@ export default class SnapTarget extends Target {
     if (this.isUseTemplateApp) {
       args.push("--template-url", `electron4:${snapArch}`)
     }
+
+    if (options.compression != null) {
+      args.push("--compression", options.compression)
+    }
     await executeAppBuilder(args)
 
     await packager.info.callArtifactBuildCompleted({
diff --git a/packages/builder-util/package.json b/packages/builder-util/package.json
index 178309cc..8e2166fa 100644
--- a/packages/builder-util/package.json
+++ b/packages/builder-util/package.json
@@ -18,7 +18,7 @@
     "7zip-bin": "~5.1.1",
     "@types/debug": "^4.1.6",
     "@types/fs-extra": "^9.0.11",
-    "app-builder-bin": "3.7.1",
+    "app-builder-bin": "4.1.0",
     "bluebird-lst": "^1.0.9",
     "builder-util-runtime": "workspace:*",
     "chalk": "^4.1.1",
diff --git a/pnpm-lock.yaml b/pnpm-lock.yaml
index cf64bfb2..d151bb3b 100644
--- a/pnpm-lock.yaml
+++ b/pnpm-lock.yaml
@@ -171,7 +171,7 @@ importers:
       '@types/js-yaml': 4.0.2
       '@types/source-map-support': 0.5.4
       7zip-bin: ~5.1.1
-      app-builder-bin: 3.7.1
+      app-builder-bin: 4.1.0
       bluebird-lst: ^1.0.9
       builder-util-runtime: workspace:*
       chalk: ^4.1.1
@@ -187,7 +187,7 @@ importers:
       '@types/debug': 4.1.7
       '@types/fs-extra': 9.0.12
       7zip-bin: 5.1.1
-      app-builder-bin: 3.7.1
+      app-builder-bin: 4.1.0
       bluebird-lst: 1.0.9
       builder-util-runtime: link:../builder-util-runtime
       chalk: 4.1.2
@@ -2899,8 +2899,8 @@ packages:
       normalize-path: 3.0.0
       picomatch: 2.3.0
 
-  /app-builder-bin/3.7.1:
-    resolution: {integrity: sha512-ql93vEUq6WsstGXD+SBLSIQw6SNnhbDEM0swzgugytMxLp3rT24Ag/jcC80ZHxiPRTdew1niuR7P3/FCrDqIjw==}
+  /app-builder-bin/4.1.0:
+    resolution: {integrity: sha512-rbdMe0sIVE95cpYqMQh4IFqhTDdB8LkKlTRcbO/Y3QleRYoIePejIbX774IYomYYzZbJfJuX7pLRiGvSdIxIYA==}
     dev: false
 
   /archiver-utils/2.1.0:

```

File: packages/app-builder-lib/scheme.json
Comment (ppd) [inline packages/app-builder-lib/scheme.json:5237]: The default should probably be lzo. There is mostly just one reason for xz: file size. The abysmal decompression speed is absolutely terrible for the kind of applications that are usually built with Electron.

The kde-neon extension wants to enable lzo by default in the future: https://github.com/snapcore/snapcraft/pull/3595

Comment (om26er) [reply packages/app-builder-lib/scheme.json:5237]: @ppd can you propose the change to my PR, I'll merge that in

Comment (ppd) [reply packages/app-builder-lib/scheme.json:5237]: @om26er https://github.com/om26er/electron-builder/pull/1
Contains the changes from the patch above and the lzo default.

**Comment by ppd** [comment]
@om26er there are some additional changes @ https://github.com/om26er/electron-builder/pull/2

> Actually default to lzo by including it in the snapcraft template.
> Also construct the builder argument from the descriptor & not from the option. We would ignore the compression option from the default template otherwise.
> 
> Also included: two rather superficial tests. I think the jest snapshots need to be updated, still.

**Comment by ppd** [comment]
@develar I don't know what to do about this:

> /root/project/node_modules/.pnpm/app-builder-bin@4.1.0/node_modules/app-builder-bin/linux/x64/app-builder exited with code ERR_ELECTRON_BUILDER_CANNOT_EXECUTE

The tests run without a problem locally. 

@om26er Also, the jest snapshots need to be regenerated: https://github.com/om26er/electron-builder/pull/3

**Comment by ppd** [comment]
@om26er A changeset is needed, I guess. What do you think?

Something like

```markdown
---
"app-builder-lib": minor
---

Default to LZO compression for snap packages. This greatly improves cold startup performance (https://snapcraft.io/blog/why-lzo-was-chosen-as-the-new-compression-method). LZO has already been adopted by most desktop-oriented snaps outside of the Electron realm.

For the rare case where developers prefer a smaller file size (XZ) to vastly improved decompression performance (LZO),
provide an option to override the default compression method.

Consumers do not need to update their configuration unless they specifically want to stick to XZ compression.
```

The dependency update gets picked up automatically by the changeset  bot, I presume?

**Comment by om26er** [comment]
Thanks @ppd for your PRs.

@maintainers does it look good now ?

**Comment by mmaietta** [comment]
Great work!
Re: updating:
>-      app-builder-bin: 3.7.1
+      app-builder-bin: 4.1.0

Is this a breaking electron-builder upgrade? I see this comment for app-builder, but the `.changeset` file shows as a `minor`
> I published app-builder-bin@4.0.0. The reason for breaking change — Go 1.17 requires macOS 10.13 High Sierra or later. I decided that it is ok.

**Comment by ppd** [comment]
@mmaietta Making it a major version increase would make sense from the system compatibility POV. I think I thought about it in the API/config sense when I wrote the changeset; it's a non-breaking change from that perspective.

So let's make it a major changeset then?! @om26er

**Comment by om26er** [comment]
> So let's make it a major changeset then?! @om26er

yeah, sounds good to me

**Comment by ppd** [comment]
@om26er https://github.com/om26er/electron-builder/pull/5

**Comment by mmaietta** [comment]
Great to see the tests passing!

Hmmm re: 
```
app-builder-bin@4.1.0/node_modules/app-builder-bin/linux/x64/app-builder exited with code ERR_ELECTRON_BUILDER_CANNOT_EXECUTE
```
I'm wondering if GO version needs to be updated in the docker images as well. Could someone please check that? I only have access to my hackintosh during these holidays, so I can't run anything related to docker atm.

@develar, can you please grant me access to our dockerhub account so I can perform this maintenance?

**Comment by ppd** [comment]
@mmaietta There's this error in the logs:

`error while loading shared libraries: liblzo2.so.2`

E.g. https://app.circleci.com/pipelines/github/electron-userland/electron-builder/1097/workflows/fe032184-a970-40f1-a9ee-bceaa7f9ea6c/jobs/4847/parallel-runs/2?filterBy=FAILED&invite=true#step-106-47

That probably needs to be added to the docker image.

**Comment by mmaietta** [comment]
So it seems I can't just add it to the docker image locally as it, unsurprisingly, doesn't build on my M1 mac. I can get it to build via GH Actions though.

If you're willing to try it out, as I'm unable to test locally, I think the way to fix this would be to add `liblzo2-dev` to the base Dockerfile
https://github.com/electron-userland/electron-builder/blob/cd79c53828759cf19cd361a48ef6fd57fff0e2f1/docker/base/Dockerfile#L13-L16
Then for testing the image locally, I think this should work? (Haven't worked much with our Docker setup yet)
```
sh docker/build.sh
TEST_FILES="snapTest" pnpm test-linux
```

**Comment by ppd** [comment]
@mmaietta 

Looks good:
```shell
Test Suites: 1 passed, 1 total
Tests:       13 passed, 13 total
Snapshots:   33 passed, 33 total
Time:        38.689 s
Ran all test suites matching /snapTest\.ts$/i.
```

with https://github.com/om26er/electron-builder/pull/6 @om26er

**Comment by mmaietta** [comment]
Can you please pull in latest `master`?
I've just merged a PR that enables testing Dockerfile changes within the CI build itself
https://github.com/electron-userland/electron-builder/blob/d835aba702caa3efc7788c7c1c0b1e11638ee4c1/.github/workflows/test.yaml#L31-L39

After that, we can test the full Snap compression flow and we should be good to go from there!

**Comment by ppd** [comment]
> Can you please pull in latest `master`?

@om26er

**Comment by om26er** [comment]
> > Can you please pull in latest `master`?
> 
> @om26er

done, lets see how it goes

**Comment by ppd** [comment]
am I blind, or does Test not run the snap tests?

**Comment by mmaietta** [comment]
I was blind. Didn't realize there were circleci tests running on an electron-builder docker image. pnpm support on circleci is pretty meh too, so I'm seeing if I can just migrate the tests to Github Actions off of circleci

**Comment by mmaietta** [comment]
I've migrated all tests to GH Actions from TravisCI and CircleCI in https://github.com/electron-userland/electron-builder/pull/6541. The TravisCI tests weren't even running 🤯 

Can you please pull in latest `master` branch?

**Comment by om26er** [comment]
> Can you please pull in latest `master` branch?

done, let's see how it goes

**Review by mmaietta — APPROVED** [review]

**Comment by mmaietta** [comment]
Okay okay, I think I've stabilized all tests on master now, and it should properly trigger on your PR as well now.

Please pull in latest master again (apologies on the delays here, thanks for your patience 😄 )

**Comment by mmaietta** [comment]
Great work guys! Really appreciate the contribution

**Comment by om26er** [comment]
Thanks @mmaietta for the review and thank you @ppd for taking this PR forward

**Comment by ppd** [comment]
Thank you, @om26er for the heavy lifting, and @mmaietta for taking care of this PR. I hope someone will step up and integrate core20 (and core22) support soon, too.

**Comment by mmaietta** [comment]
Published in `23.0.0-alpha.0` and just deployed new versions of the docker images. Enjoy!

**Comment by julian-alarcon** [comment]
This was partially adding the feature, but there are still issues: https://github.com/electron-userland/electron-builder/issues/7013

**Pull request merged**
